### ISSUE TITLE
CompoundGeometry.from_points with control_points=None

### ISSUE THREAD

**Issue by normanrichardson** [robbievanleeuwen/section-properties#113]
**Describe the bug**
When using `CompoundGeometry.from_points` with control_points=None, the class method fails. 

**To Reproduce**
```python
import sectionproperties.pre.sections as sections
from sectionproperties.analysis.cross_section import Section

points = [[0,0], [0,1], [1,1], [1,0], [1,0.25], [1,0.75], [2,0.75], [2,0.25] ]
facets = [[0,1], [1,2], [2,5],[5,4], [4,3], [3,0], [4,5], [5,6], [6,7], [7,4]]
holes = []

geometry = sections.CompoundGeometry.from_points(points, facets, holes)
```
```
TypeError: 'NoneType' object is not iterable
````
originating at [here](https://github.com/robbievanleeuwen/section-properties/blob/b661bb093497d819bcbdf0e8be3028872c202df3/sectionproperties/pre/sections.py#L1177)

**Expected behaviour**
With control_points=None and material=None, I would expect Shapely representative points to be generated.

**Comment by connorferster** [comment]
@normanrichardson Agreed on the desired behaviour.

**Comment by connorferster** [comment]
@normanrichardson, would be interested to hear your thoughts on this change to `CompoundGeometry.from_points()`

**Comment by normanrichardson** [comment]
@connorferster I had a chance to have a look. It seems that the fixes approach is to require control points for all regions. To elaborate on what I thought the expected behavior would be:

Working off #117 and the above example (modified) I'm getting:
```python
import sectionproperties.pre.sections as sections
from sectionproperties.analysis.cross_section import Section

points = [[0,0], [0,1], [1,1], [1,0], [1,0.25], [1,0.75], [2,0.75], [2,0.25]]
facets = [[0,1], [1,2], [2,3],[3,0], [4,5], [5,6], [6,7], [7,4]]
holes = []
control_points = None

geometry = sections.CompoundGeometry.from_points(points, facets, holes, control_points)
geometry.plot_geometry(labels=["points", "facets"])
```
```
ValueError: The number of exterior regions (2) does not match the number of control_points given (0).
```
When I contrast this behavior to a `Geometry` example:
```python
import sectionproperties.pre.sections as sections
from sectionproperties.analysis.cross_section import Section

points = [[0,0], [0,1], [1,1], [1,0]]
facets = [[0,1], [1,2], [2,3],[3,0]]
holes = []
control_points = None

geometry = sections.Geometry.from_points(points, facets, holes, control_points)
geometry.plot_geometry(labels=["points", "facets"])
```
![Figure_0](https://user-images.githubusercontent.com/14271499/143090397-a93b15d7-59c7-4004-83eb-66178650481b.png)

I would have expected to have a comparable behavior, that: "regions not claimed by control points or holes get assigned control points (via a representative point)"). That said my anticipated behavior may have its own issues.

**Comment by normanrichardson** [comment]
Something else that I noticed when experimenting #117: when interiors are removed [here](https://github.com/robbievanleeuwen/section-properties/blob/233ad21b6692bbe334a3501e469547f14facd345/sectionproperties/pre/sections.py#L1204-L1211) a `Geometry` is subtracted from a `CompoundGeometry`, which throws a ValueError:
```
ValueError: Cannot perform 'difference' on these two objects: <sectionproperties.pre.sections.CompoundGeometry object at 0x7fdde249a550> - <sectionproperties.pre.sections.Geometry object at 0x7fddc8cd4670>
```
Something, like [this](https://github.com/robbievanleeuwen/section-properties/blob/233ad21b6692bbe334a3501e469547f14facd345/sectionproperties/pre/sections.py#L1213-L1214) would correct this.

**Comment by connorferster** [comment]
@normanrichardson and @robbievanleeuwen 

I have been thinking on this further today. I think it is reasonable to require the user to manually input each control point. The alternative is that sectionproperties would have to implement a robust polygon nesting algorithm to determine order of precedence when it comes to overlapping or nested geometries, as shown in my notebook below:

![image](https://user-images.githubusercontent.com/38441489/146983863-ed56e82a-a4d4-4fc7-ad41-2efc4c5b369f.jpeg)

**Comment by connorferster** [comment]
Aside from this issue, I am realizing when it comes to multiple nested geometries, my `.from_points()` process is not robust enough. In fact, multiple nested geometries with holes seems to be tricky business. Will report back when I get it figured out.

**Comment by connorferster** [comment]
Is it even possible to unambiguously define a multiple nested geometry with holes through just points, facets, control_points, and holes? How would this be done in v1? Would you do a similar approach to what I showed above? Where each region (polygon - hole) would be defined as a `CustomSection` and then all three boundaries would be then combined in a `MergedSection`? 

The interesting thing is that the multiple nested geometry meshes in Triangle beautifully and Triangle is only getting points, facets, control_points, and holes.

```python
sq1 = sections.rectangular_section(100, 100).align_center()
sq2 = sections.rectangular_section(75, 75).align_center()
sq3 = sections.rectangular_section(50, 50).align_center()
hole = sections.rectangular_section(25, 25).align_center()
compound = (sq1 - sq2) + (sq2 - sq3) + (sq3 - hole)
compound.create_mesh([10, 5, 2.5])
c_sec = Section(compound)
c_sec.plot_mesh()
```

![image](https://user-images.githubusercontent.com/38441489/147044886-bc52faed-16e3-4c95-b083-7080c1aef6bb.png)

**Comment by robbievanleeuwen** [comment]
Hi all,

Agreed that while *sectionproperties* IS becoming a better pre-processor with the excellent addition of shapely, it is still largely an analysis tool and I don't think we should should be solving extremely complex geometry problems for the sake of pre-processing, which in fact may not be solveable for all arbitrary geometries!

While I agree that in an ideal world, `control_points` would not be a required input, it seems like this is extremely difficult to implement. @connorferster I think you've hit the nail on the head in that `.from_points()` users will have an exact idea of what the geometry looks like and will be able to specify `control_points` easily.

I'm not sure if I'm missing something, but in v1 I think the above geometry could just be created from one single `CustomSection`. There are 16 points & 16 facets defined, as well as 3 control points & one hole. The fact that the facets do not form one continous loop or 'linear ring' does not concern triangle and the mesh is easily generated. Might be missing something here though?

@connorferster as the chief shapely<->sectionproperties architect, if you are happy with the current implementation I'm happy to close this one, but an interesting discussion to have nonetheless!

Robbie

**Comment by connorferster** [comment]
@robbievanleeuwen 

Thanks for that. You gave me the insight I needed to solve this problem. Through this exploration, I have realized that my existing implementation is not robust enough but I could not figure out a path forward.

Here's how it looks to me now:

* shapely has a concept of z-ordering which means holes need to be added to each separate solid if they are to appear nested (thus, three regions + three holes, like a matryoshka). In other words, holes are relative to the one region they are immediately "over".
* Triangle has no concept of z-ordering. Holes are always "absolute" and "punch through" all "layered" regions.

With this clarity, I hope to be able to figure this out!

**Comment by connorferster** [comment]
Baaaammm! Nailed it.

Will write some new tests for `CompoundGeometry.from_points()` and add a new section to Advanced Geometry in the docs before submitting the PR but here is the more robust `.from_points()` algorithm at work:

![image](https://user-images.githubusercontent.com/38441489/147164898-0ba34432-ba04-4f36-b825-3d5009e40982.png)

**Comment by normanrichardson** [comment]
@connorferster The "possible scenarios" you showed, are the exact complexity issue I was beginning to see myself, by allowing control_points=None in CompoundGeometry.

My intent behind this issue, was that I feel that for the user it would be ambiguous to allow control_points=None in Geometry and not allow control_points=None in CompoundGeometry. I feel this way about the behavior of the two functions in general, what the one does the other should do. That said, fundamentally there is nothing wrong with allowing different behaviors, and I see many viable solutions here:
1. CompoundGeometry.from_points is different from Geometry.from_points. We allow control_points=None in Geometry and document it nicely.
2. CompoundGeometry.from_points and Geometry.from_points both don't allow control_points=None.
3. CompoundGeometry.from_points and Geometry.from_points both allow control_points=None.
4. We deprecate CompoundGeometry.from_points and Geometry.from_points

I think 4 is the most difficult solution as we then have to look hard at the whole library and consider other deprecations.

I see 3 as a doable but perhaps difficult solution. I also agree with Robbie's feelings questioning scope. I can see another issue looming: a user will put in one set of data (say facets), and different data could come out! I think this would happen in "possible scenarios" 4 (two overlapping rectangles). A facet list goes in, and all of a sudden the facet list is now modified with additional entries, etc. I think a user may find this confusing. If this is not clear, let me know and I can make an example.

I see 1 and 2 as more or less equal, easy enough to implement. 2 would mean the user would have to be 100% explicit about the control_points, while reducing convenience.

If I'm missing solutions let me know.

**Comment by connorferster** [comment]
@normanrichardson I agree on your four possibilities and am coming around to the idea of mandatory entry of both facets and control points for both Geometry and CompoundGeometry.

The way I am thinking about .from_points(), currently, is that this is the "Triangle interface" which is the only interface V1 supports. Thus, the Triangle interface should be consistent across both classes.

Speaking against, the desire is to reduce the entry of redundant information wherever possible.

**Comment by connorferster** [comment]
Incidentally, I think the new .from_points algorithm now allows for entry of all the notebook examples, especially once I realized that there is no z-ordering in the Triangle interface and that entering an "overlapping" (ambiguous) geometry via .from_points would be a poor use if time. So, that's good news!

**Comment by connorferster** [comment]
@normanrichardson Design question for you:

`Geometry.from_points()` accepts a control point as a `List[float, float]`, e.g. `[0., 0.]` because a `Geometry` assumes only a single control point. `CompoundGeometry.from_points()`, conversely, accepts multiple control points as a `List[List[float, float]]`, e.g. `[[0., 0.,]]` but does not require multiple points to be entered.

Keeping on the consistency train, do you think `Geometry.from_points()` should accept a single control point as a list of list of float, also?

Pros: 
* It would make error checking of `len(control_points)` easier if the same format of data is expected for both
* It would be consistent with `.create_mesh()` where the modeller is expected to enter a list of mesh sizes even for a `Geometry`

Cons:
* It may be surprising to the modeller that they cannot enter a single control point as an `[x, y]` coordinate directly

**Comment by robbievanleeuwen** [comment]
Addressed with #133

**Comment by normanrichardson** [comment]
@connorferster this is a really good point. If we think of the Geometry.from_points and CompoundGeometry.from_points as being ways to define the Triangle data (that then go and construct the shapely data), then yes that seems consistent. I can see how that is clunky though.

This brings up something I have been feeling that maybe you have already considered and have already discovered the issues with: Could v2 (at some later stage) at the Geometry/CompoundGeometry level break away from Triangle, ie. just make it a Shapely object (in attr geom) + metadata attributes (e.g. material/materials). Then at create_mesh time, it would construct the triangle attributes/data. It seems like it would lower the overhead of maintaining both the Triangle data and the Shapely data.
